trin-core/src/portalnet/overlay_service.rs

```diff
@@ -211,28 +384,56 @@ impl OverlayService {
         *self.data_radius
     }
 
+    /// Maintains the routing table.
+    ///
+    /// Consumes previously applied pending entries from the `KBucketsTable`. An `AppliedPending`
+    /// result is recorded when a pending bucket entry replaces a disconnected entry in the
+    /// respective bucket.
+    async fn bucket_maintenance_poll(
+        protocol: ProtocolId,
+        kbuckets: &Arc<RwLock<KBucketsTable<NodeId, Node>>>,
+    ) {
+        // Drain applied pending entries from the routing table.
+        if let Some(entry) = kbuckets.write_with_warn().await.take_applied_pending() {
+            debug!(
+                "[{:?}] Node {:?} inserted and node {:?} evicted",
+                protocol,
+                entry.inserted.into_preimage(),
+                entry.evicted.map(|n| n.key.into_preimage())
+            );
+        }
+    }
```

Comment:
Good callout! I don't think we want to sleep in our code though. I think what we want is something similar to what [`discv5` does](https://github.com/sigp/discv5/blob/fdbb80432419578eebadf27c7d6fb098975f5de6/src/service.rs#L1355), and implement the `Future` trait for this function. This is what I tried to implement initially, but I ran into some issues that I deferred at the time. Time to take another attempt at this though.

Reply:
So to fix this, I moved `Overlay` and `OverlayService` to use `parking_lot::RwLock` instead of `tokio::sync::RwLock`. This allowed me to make a lot of previously async functions non-async. This change was necessary to implement the `future::poll_fn` linked above. Otherwise we cannot call `await` in the `future::poll_fn` closure. My understanding is that these non-async `RwLock` instances are okay as long as we do not hold the lock across `await` points in async contexts.

I also replaced `write_with_warn` with the standard `write`. This is because those lock utilities are implemented for `tokio::sync::RwLock`.

If this represents a significant enough change to merit its own PR, then I can raise one.